// @spotifly/auth-token/cli/dist/spotifly-auth-token-cli.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var authorize = require('../../dist/authorize-c5d0f43d.cjs.prod.js');
require('@spotifly/utils');
require('node-fetch');
require('@eegli/tinyparse');
require('http');

var _package = {
	name: "@spotifly/auth-token",
	version: "1.0.0",
	description: "A simple utility for retrieving Spotify authentication and refresh tokens. Supports programmatic and CLI use",
	keywords: [
		"spotify",
		"auth",
		"token",
		"refresh token",
		"authorization",
		"authentication",
		"cli",
		"typescript"
	],
	homepage: "https://spotifly.nougat.dev/docs/packages/auth-token",
	repository: {
		type: "git",
		url: "https://github.com/eegli/spotifly"
	},
	license: "MIT",
	author: {
		name: "Eric Egli",
		email: "lx180777+npm@gmail.com"
	},
	main: "dist/spotifly-auth-token.cjs.js",
	module: "dist/spotifly-auth-token.esm.js",
	files: [
		"dist",
		"cli"
	],
	dependencies: {
		"@eegli/tinyparse": "^0.7.0",
		"@spotifly/utils": "workspace:^",
		"node-fetch": "2.6.7"
	},
	devDependencies: {
		"@types/fs-extra": "^9.0.13",
		"@types/node-fetch": "^2.6.2"
	},
	engines: {
		node: ">=14"
	},
	publishConfig: {
		directory: "_release/package"
	},
	preconstruct: {
		entrypoints: [
			"index.ts",
			"cli.ts"
		]
	}
};

const callback = async argv => {
  const config = await authorize.parse(argv);
  await authorize.authorize(config);
};

exports.help = authorize.help;
exports.callback = callback;
exports.pkg = _package;
